src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java

```diff
@@ -609,95 +610,89 @@ private synchronized String getHeadsAsString() {
         } else if (myHeads.size() == 0) {
             return "empty-heads";
         } else {
-            return myHeads.entrySet().stream().map(h->h.getKey()+"::"+h.getValue()).collect(Collectors.joining(","));
+            return myHeads.entrySet().stream().map(h -> h.getKey() + "::" + h.getValue()).collect(Collectors.joining(","));
         }
-
     }
 
-    /** Handle failures where we can't load heads. */
-    private void rebuildEmptyGraph() {
-        synchronized (this) {
-            // something went catastrophically wrong and there's no live head. fake one
-            LOGGER.log(Level.WARNING, "Failed to load pipeline heads, so faking some up for execution " + this.toString());
-            if (this.startNodes == null) {
-                this.startNodes = new Stack<BlockStartNode>();
+    /**
+     * In the event we're missing FlowNodes, fail-fast and create some mockup FlowNodes so we can continue.
+     * This avoids nulling out all of the execution's data
+     * Bypasses {@link #croak(Throwable)} and {@link #onProgramEnd(Outcome)} to guarantee a clean path.
+     */
+    @GuardedBy("this")
+    void createPlaceholderNodes(Throwable failureReason) throws Exception {
+        this.done = true;
+
+        if (this.owner != null) {
+            Queue.Executable ex = owner.getExecutable();
+            if (ex instanceof Run) {
+                Result res = ((Run)ex).getResult();
+                setResult(res != null ? res : Result.FAILURE);
```

Comment:
This seems weird. Normally `null` is a synonym for `SUCCESS`. Why would this not be unconditionally `FAILURE`?